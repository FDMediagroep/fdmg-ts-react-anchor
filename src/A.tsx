import * as React from 'react';
import {MouseEvent} from 'react';

export interface Props {
    ariaLabel?: string;
    children?: JSX.Element[]|HTMLElement[]|string|number;
    className?: string;
    href?: string;
    onClick?: (event: MouseEvent<HTMLAnchorElement>) => void;
    target?: string;
    tabIndex?: number;
    title?: string;
}

/**
 * A compononent renders a heading
 */
export default class A extends React.Component<Props, any> {
    public state: any;
    public props: any;

    constructor(props: Props) {
        super(props);
        this.props = props;
        console.info('A component', props);
    }

    /**
     * Returns the ARIA-label text by following priority:
     * 1. ariaLabel property
     * 2. title property
     * 3. innerText
     *
     * @returns {string}
     */
    getAriaLabel(): string {
        let ariaLabel;
        if(this.props.ariaLabel) {
            ariaLabel = this.props.ariaLabel;
        } else if(this.props.title) {
            ariaLabel = this.props.title;
        } else if(this.props.children && typeof(this.props.children) === 'string') {
            ariaLabel = this.props.children;
        }
        return ariaLabel;
    }

    render() {
        return (
            <a href={this.props.href} className={this.props.className} aria-label={this.getAriaLabel()}
               target={this.props.target} tabIndex={this.props.tabIndex?this.props.tabIndex:0}
               title={this.props.title} onClick={this.props.onClick}>
                {this.props.children}
            </a>
        );
    }
}
