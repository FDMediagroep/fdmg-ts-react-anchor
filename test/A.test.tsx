import * as React from 'react';
import * as renderer from 'react-test-renderer';
import A from '../src/A';

console.info = function() {};

test('A renders correctly', () => {
    let component = renderer.create(
        <A/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A><span>testing 123</span><div>testing456</div></A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('A renders correctly with href', () => {
    const component = renderer.create(
        <A href="https://fd.nl">test anchor</A>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('A renders correctly with CSS class', () => {
    const component = renderer.create(
        <A className={'css-class-name'}>test anchor</A>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('A renders correctly with tabIndex', () => {
    let component = renderer.create(
        <A tabIndex="2">test anchor</A>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A tabIndex="-1">test anchor</A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('A renders correctly with ARIA-label', () => {
    let component = renderer.create(
        <A>anchor text as ARIA-label</A>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A ariaLabel="ARIA-label">anchor text, not ARIA</A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A title="title shown as ARIA-label">anchor text, not ARIA</A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A title="this should not be shown as ARIA-label" ariaLabel="ariaLabel shown as ARIA-label">anchor text, not ARIA</A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A title="title shown as ARIA-label"><span>testing 123</span><div>testing456</div></A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <A ariaLabel="ariaLabel shown as ARIA-label"><span>testing 123</span><div>testing456</div></A>
    );
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
