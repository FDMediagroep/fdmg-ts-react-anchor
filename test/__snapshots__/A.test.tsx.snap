// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor renders correctly 1`] = `
<a
  tabIndex={0}
/>
`;

exports[`Anchor renders correctly 2`] = `
<a
  tabIndex={0}
>
  <span>
    testing 123
  </span>
  <div>
    testing456
  </div>
</a>
`;

exports[`Anchor renders correctly with ARIA-label 1`] = `
<a
  aria-label="anchor text as ARIA-label"
  tabIndex={0}
>
  anchor text as ARIA-label
</a>
`;

exports[`Anchor renders correctly with ARIA-label 2`] = `
<a
  aria-label="ARIA-label"
  tabIndex={0}
>
  anchor text, not ARIA
</a>
`;

exports[`Anchor renders correctly with ARIA-label 3`] = `
<a
  aria-label="title shown as ARIA-label"
  tabIndex={0}
  title="title shown as ARIA-label"
>
  anchor text, not ARIA
</a>
`;

exports[`Anchor renders correctly with ARIA-label 4`] = `
<a
  aria-label="ariaLabel shown as ARIA-label"
  tabIndex={0}
  title="this should not be shown as ARIA-label"
>
  anchor text, not ARIA
</a>
`;

exports[`Anchor renders correctly with ARIA-label 5`] = `
<a
  aria-label="title shown as ARIA-label"
  tabIndex={0}
  title="title shown as ARIA-label"
>
  <span>
    testing 123
  </span>
  <div>
    testing456
  </div>
</a>
`;

exports[`Anchor renders correctly with ARIA-label 6`] = `
<a
  aria-label="ariaLabel shown as ARIA-label"
  tabIndex={0}
>
  <span>
    testing 123
  </span>
  <div>
    testing456
  </div>
</a>
`;

exports[`Anchor renders correctly with CSS class 1`] = `
<a
  aria-label="test anchor"
  className="css-class-name"
  tabIndex={0}
>
  test anchor
</a>
`;

exports[`Anchor renders correctly with href 1`] = `
<a
  aria-label="test anchor"
  href="https://fd.nl"
  tabIndex={0}
>
  test anchor
</a>
`;

exports[`Anchor renders correctly with tabIndex 1`] = `
<a
  aria-label="test anchor"
  tabIndex="2"
>
  test anchor
</a>
`;

exports[`Anchor renders correctly with tabIndex 2`] = `
<a
  aria-label="test anchor"
  tabIndex="-1"
>
  test anchor
</a>
`;
